src/storage/src/source/generator/key_value.rs

```diff
@@ -0,0 +1,584 @@
+struct UpdateProducer {
+    // The key iterator for the partition.
+    pi: PartitionKeyIterator,
+    // The batch size.
+    batch_size: u64,
+    // The next offset to produce updates at.
+    next_offset: u64,
+    // The source-level seed for the rng.
+    seed: u64,
+    // The number of offsets we expect to be part of the `quick_rounds`.
+    expected_quick_offsets: u64,
+    // Whether to include the offset or not.
+    include_offset: bool,
+}
+
+impl UpdateProducer {
+    fn new(
+        partition: u64,
+        partitions: u64,
+        keys: u64,
+        snapshot_rounds: u64,
+        quick_rounds: u64,
+        batch_size: u64,
+        seed: u64,
+        next_offset: u64,
+        include_offset: bool,
+    ) -> Self {
+        // Each snapshot _round_ is associated with an offset, starting at 0. Therefore
+        // the `next_offset` - `snapshot_rounds` is the index of the next non-snapshot update
+        // batches we must produce. The start key for the that batch is that index
+        // multiplied by the batch size and the number of partitions.
+        let start_key =
+            (((next_offset - snapshot_rounds) * batch_size * partitions) + partition) % keys;
+
+        // We expect to emit 1 batch per offset. A round is emitting the full set of keys (in the
+        // partition. The number of keys divided by the batch size (and partitioned, so divided by
+        // the partition count) is the number of offsets in each _round_. We also add the number of
+        // snapshot rounds, which is also the number of offsets in the snapshot.
+        let expected_quick_offsets =
+            ((keys / partitions / batch_size) * quick_rounds) + snapshot_rounds;
+
+        let pi = PartitionKeyIterator::new(partition, partitions, keys, start_key);
+        UpdateProducer {
+            pi,
+            batch_size,
+            next_offset,
+            seed,
+            expected_quick_offsets,
+            include_offset,
+        }
+    }
+
+    /// If this partition is done producing quick updates.
+    fn finished_quick(&self) -> bool {
+        self.next_offset >= self.expected_quick_offsets
+    }
+
+    /// Produce a batch of message into `buffer`, of size `batch_size`. Advances the current
+    /// batch and round counter. Returns the frontier after the batch.
+    fn produce_batch(
+        &mut self,
+        buffer: &mut Vec<(
+            (usize, Result<SourceMessage, SourceReaderError>),
+            MzOffset,
+            Diff,
+        )>,
+        value_buffer: &mut Vec<u8>,
+    ) -> u64 {
+        // Consistently seeded with the source see, the offset, and the
+        // partition.
+        let mut seed = [0; 32];
+        seed[0..8].copy_from_slice(&self.seed.to_le_bytes());
+        seed[8..16].copy_from_slice(&self.next_offset.to_le_bytes());
+        seed[16..24].copy_from_slice(&self.pi.partition.to_le_bytes());
+        let mut rng = StdRng::from_seed(seed);
```

Comment:
This should be extracted in its own function since it's copy pasted in two places currently and it's critical for the source to produce a sensible collection